File: repos/Textractor-cmd/host/hookcode.cpp
#include "hookcode.h"
#include "module.h"

namespace
{
	std::optional<HookParam> ParseRCode(std::wstring RCode)
	{
		std::wsmatch match;
		HookParam hp = {};
		hp.type |= DIRECT_READ;

		// {S|Q|V|M}
		switch (RCode[0])
		{
		case L'S':
			break;
		case L'Q':
			hp.type |= USING_UNICODE;
			break;
		case L'V':
			hp.type |= USING_UTF8;
			break;
		case L'M':
			hp.type |= USING_UNICODE | HEX_DUMP;
			break;
		default:
			return {};
		}
		RCode.erase(0, 1);

		// [null_length<]
		if (std::regex_search(RCode, match, std::wregex(L"^([0-9]+)<")))
		{
			hp.null_length = std::stoi(match[1]);
			RCode.erase(0, match[0].length());
		}

		// [codepage#]
		if (std::regex_search(RCode, match, std::wregex(L"^([0-9]+)#")))
		{
			hp.codepage = std::stoi(match[1]);
			RCode.erase(0, match[0].length());
		}

		// @addr
		if (!std::regex_match(RCode, match, std::wregex(L"@([[:xdigit:]]+)"))) return {};
		hp.address = std::stoull(match[1], nullptr, 16);
		return hp;
	}

	std::optional<HookParam> ParseHCode(std::wstring HCode)
	{
		std::wsmatch match;
		HookParam hp = {};

		// {A|B|W|H|S|Q|V|M}
		switch (HCode[0])
		{
		case L'A':
			hp.type |= BIG_ENDIAN;
			hp.length_offset = 1;
			break;
		case L'B':
			hp.length_offset = 1;
			break;
		case L'W':
			hp.type |= USING_UNICODE;
			hp.length_offset = 1;
			break;
		case L'H':
			hp.type |= USING_UNICODE | HEX_DUMP;
			hp.length_offset = 1;
			break;
		case L'S':
			hp.type |= USING_STRING;
			break;
		case L'Q':
			hp.type |= USING_STRING | USING_UNICODE;
			break;
		case L'V':
			hp.type |= USING_STRING | USING_UTF8;
			break;
		case L'M':
			hp.type |= USING_STRING | USING_UNICODE | HEX_DUMP;
			break;
		default:
			return {};
		}
		HCode.erase(0, 1);

		if (hp.type & USING_STRING)
		{
			if (HCode[0] == L'F')
			{
				hp.type |= FULL_STRING;
				HCode.erase(0, 1);
			}

			// [null_length<]
			if (std::regex_search(HCode, match, std::wregex(L"^([0-9]+)<")))
			{
				hp.null_length = std::stoi(match[1]);
				HCode.erase(0, match[0].length());
			}
		}

		// [N]
		if (HCode[0] == L'N')
		{
			hp.type |= NO_CONTEXT;
			HCode.erase(0, 1);
		}

		// [codepage#]
		if (std::regex_search(HCode, match, std::wregex(L"^([0-9]+)#")))
		{
			hp.codepage = std::stoi(match[1]);
			HCode.erase(0, match[0].length());
		}

		// [padding+]
		if (std::regex_search(HCode, match, std::wregex(L"^([[:xdigit:]]+)\\+")))
		{
			hp.padding = std::stoull(match[1], nullptr, 16);
			HCode.erase(0, match[0].length());
		}

		auto ConsumeHexInt = [&HCode]
		{
			size_t size = 0;
			int value = 0;
			try { value = std::stoi(HCode, &size, 16); } catch (std::invalid_argument) {}
			HCode.erase(0, size);
			return value;
		};

		// data_offset
		hp.offset = ConsumeHexInt();

		// [*deref_offset1]
		if (HCode[0] == L'*')
		{
			hp.type |= DATA_INDIRECT;
			HCode.erase(0, 1);
			hp.index = ConsumeHexInt();
		}

		// [:split_offset[*deref_offset2]]
		if (HCode[0] == L':')
		{
			hp.type |= USING_SPLIT;
			HCode.erase(0, 1);
			hp.split = ConsumeHexInt();

			if (HCode[0] == L'*')
			{
				hp.type |= SPLIT_INDIRECT;
				HCode.erase(0, 1);
				hp.split_index = ConsumeHexInt();
			}
		}

		// @addr[:module[:func]]
		if (!std::regex_match(HCode, match, std::wregex(L"^@([[:xdigit:]]+)(:.+?)?(:.+)?"))) return {};
		hp.address = std::stoull(match[1], nullptr, 16);
		if (match[2].matched)
		{
			hp.type |= MODULE_OFFSET;
			wcsncpy_s(hp.module, match[2].str().erase(0, 1).c_str(), MAX_MODULE_SIZE - 1);
		}
		if (match[3].matched)
		{
			hp.type |= FUNCTION_OFFSET;
			std::wstring func = match[3];
			strncpy_s(hp.function, std::string(func.begin(), func.end()).erase(0, 1).c_str(), MAX_MODULE_SIZE - 1);
		}

		// ITH has registers offset by 4 vs AGTH: need this to correct
		if (hp.offset < 0) hp.offset -= 4;
		if (hp.split < 0) hp.split -= 4;

		return hp;
	}

	std::wstring HexString(int64_t num)
	{
		if (num < 0) return FormatString(L"-%I64X", -num);
		return FormatString(L"%I64X", num);
	}

	std::wstring GenerateRCode(HookParam hp)
	{
		std::wstring RCode = L"R";

		if (hp.type & USING_UNICODE)
		{
			if (hp.type & HEX_DUMP) RCode += L'M';
			else RCode += L'Q';
			if (hp.null_length != 0) RCode += std::to_wstring(hp.null_length) + L'<';
		}
		else
		{
			RCode += L'S';
			if (hp.null_length != 0) RCode += std::to_wstring(hp.null_length) + L'<';
			if (hp.codepage != 0) RCode += std::to_wstring(hp.codepage) + L'#';
		}

		RCode += L'@' + HexString(hp.address);

		return RCode;
	}

	std::wstring GenerateHCode(HookParam hp, DWORD processId)
	{
		std::wstring HCode = L"H";

		if (hp.type & USING_UNICODE)
		{
			if (hp.type & HEX_DUMP)
			{
				if (hp.type & USING_STRING) HCode += L'M';
				else HCode += L'H';
			}
			else
			{
				if (hp.type & USING_STRING) HCode += L'Q';
				else HCode += L'W';
			}
		}
		else
		{
			if (hp.type & USING_STRING) HCode += L'S';
			else if (hp.type & BIG_ENDIAN) HCode += L'A';
			else HCode += L'B';
		}

		if (hp.type & FULL_STRING) HCode += L'F';

		if (hp.null_length != 0) HCode += std::to_wstring(hp.null_length) + L'<';

		if (hp.type & NO_CONTEXT) HCode += L'N';
		if (hp.text_fun || hp.filter_fun || hp.hook_fun || hp.length_fun) HCode += L'X'; // no AGTH equivalent

		if (hp.codepage != 0 && !(hp.type & USING_UNICODE)) HCode += std::to_wstring(hp.codepage) + L'#';

		if (hp.padding) HCode += HexString(hp.padding) + L'+';

		if (hp.offset < 0) hp.offset += 4;
		if (hp.split < 0) hp.split += 4;

		HCode += HexString(hp.offset);
		if (hp.type & DATA_INDIRECT) HCode += L'*' + HexString(hp.index);
		if (hp.type & USING_SPLIT) HCode += L':' + HexString(hp.split);
		if (hp.type & SPLIT_INDIRECT) HCode += L'*' + HexString(hp.split_index);

		// Attempt to make the address relative
		if (processId && !(hp.type & MODULE_OFFSET))
			if (AutoHandle<> process = OpenProcess(PROCESS_VM_READ | PROCESS_QUERY_INFORMATION, FALSE, processId))
				if (MEMORY_BASIC_INFORMATION info = {}; VirtualQueryEx(process, (LPCVOID)hp.address, &info, sizeof(info)))
					if (auto moduleName = GetModuleFilename(processId, (HMODULE)info.AllocationBase))
					{
						hp.type |= MODULE_OFFSET;
						hp.address -= (uint64_t)info.AllocationBase;
						wcsncpy_s(hp.module, moduleName->c_str() + moduleName->rfind(L'\\') + 1, MAX_MODULE_SIZE - 1);
					}

		HCode += L'@' + HexString(hp.address);
		if (hp.type & MODULE_OFFSET) HCode += L':' + std::wstring(hp.module);
		if (hp.type & FUNCTION_OFFSET) HCode += L':' + StringToWideString(hp.function);

		return HCode;
	}
}

namespace HookCode
{
	std::optional<HookParam> Parse(std::wstring code)
	{
		if (code[0] == L'/') code.erase(0, 1); // legacy/AGTH compatibility
		if (code[0] == L'R') return ParseRCode(code.erase(0, 1));
		else if (code[0] == L'H') return ParseHCode(code.erase(0, 1));
		return {};
	}

	std::wstring Generate(HookParam hp, DWORD processId)
	{
		return hp.type & DIRECT_READ ? GenerateRCode(hp) : GenerateHCode(hp, processId);
	}

	TEST(
		assert(StringToWideString(u8"こんにちは") == L"こんにちは"),
		assert(HexString(-12) == L"-C"),
		assert(HexString(12) == L"C"),
		assert(Parse(L"/HQN936#-c*C:C*1C@4AA:gdi.dll:GetTextOutA")),
		assert(Parse(L"HB4@0")),
		assert(Parse(L"/RS65001#@44")),
		assert(Parse(L"HQ@4")),
		assert(!Parse(L"/RW@44")),
		assert(!Parse(L"/HWG@33"))
	);
}

File: LunaTranslator/LunaTranslator/textsource/hook/hookcode.py
import textsource.hook.define as define
import windows
from traceback import print_exc
#import define
import re,copy
USING_STRING = 0x1  # type(data) is char * or wchar_t * and has length
USING_UNICODE = 0x2  # type(data) is wchar_t or wchar_t*
BIG_ENDIAN = 0x4  # type(data) is char
DATA_INDIRECT = 0x8
USING_SPLIT = 0x10  # use ctx2 or not
SPLIT_INDIRECT = 0x20
MODULE_OFFSET = 0x40  # address is relative to module
FUNCTION_OFFSET = 0x80  # address is relative to function
USING_UTF8 = 0x100
NO_CONTEXT = 0x200
HOOK_EMPTY = 0x400
FIXING_SPLIT = 0x800
DIRECT_READ = 0x1000  # /R read code instead of classic / H hook code
FULL_STRING = 0x2000
HEX_DUMP = 0x4000
HOOK_ENGINE = 0x8000
HOOK_ADDITIONAL = 0x10000
KNOWN_UNSTABLE = 0x20000
def ConsumeHexInt(HCode):
    match = re.match(r'^([+-]?[0-9a-fA-F]+)', HCode)
    if match:
        value = int(match.group(1), 16)
        HCode = HCode[match.end():]
        return HCode,value
    else:
        return HCode,0
def Hex(st):
    return hex(st).replace('0x','').upper()
def ParseRCode(RCode) :
    hp=define.HookParam()
    hp.type |= DIRECT_READ
    if RCode[0]=='S':
        pass
    elif RCode[0] == 'Q':
        hp.type |= USING_UNICODE
    elif RCode[0]  == 'V':
        hp.type |= USING_UTF8
    elif RCode[0]  == 'M':
        hp.type |= USING_UNICODE | HEX_DUMP
    else:
        return None
    RCode = RCode[1:]
    # [null_length<]
    match = re.match(r"^([0-9]+)<", RCode)
    if match:
        hp.null_length = int(match.group(1))
        RCode = RCode[len(match.group(0)):]
    # [codepage#]
    match = re.match(r"^([0-9]+)#", RCode)
    if match:
        hp.codepage = int(match.group(1))
        RCode = RCode[len(match.group(0)):]
    # @addr
    match = re.match(r"@([0-9a-fA-F]+)", RCode)
    if not match:
        return None
    hp.address = int(match.group(1), 16)
    return hp

def ParseHCode(HCode):
    hp=define.HookParam()
    if HCode[0] == 'A':
        hp.type |= BIG_ENDIAN
        hp.length_offset = 1
    elif HCode[0] == 'B':
        hp.length_offset = 1
    elif HCode[0] == 'W':
        hp.type |= USING_UNICODE
        hp.length_offset = 1
    elif HCode[0] == 'H':
        hp.type |= USING_UNICODE | HEX_DUMP
        hp.length_offset = 1
    elif HCode[0] == 'S':
        hp.type |= USING_STRING
    elif HCode[0] == 'Q':
        hp.type |= USING_STRING | USING_UNICODE
    elif HCode[0] == 'V':
        hp.type |= USING_STRING | USING_UTF8
    elif HCode[0] == 'M':
        hp.type |= USING_STRING | USING_UNICODE | HEX_DUMP
    else:
        return None
    HCode = HCode[1:]

    if hp.type & USING_STRING:
        if HCode[0] == 'F':
            hp.type |= FULL_STRING
            HCode = HCode[1:]
        # [null_length<]
        match = re.match(r'^([0-9]+)<', HCode)
        if match:
            hp.null_length = int(match.group(1))
            HCode = HCode[match.end():]
    # [N]
    if HCode[0] == 'N':
        hp.type |= NO_CONTEXT
        HCode = HCode[1:]
    # [codepage#]
    match = re.match(r'^([0-9]+)#', HCode)
    if match:
        hp.codepage = int(match.group(1))
        HCode = HCode[match.end():]
    # [padding+]
    match = re.match(r'^([0-9a-fA-F]+)\+', HCode)
    if match:
        hp.padding = int(match.group(1), 16)
        HCode = HCode[match.end():]
    HCode,hp.offset=ConsumeHexInt(HCode)
    # [*deref_offset1]
    if HCode[0] == '*':
        hp.type |= DATA_INDIRECT
        HCode = HCode[1:]
        HCode,hp.index=ConsumeHexInt(HCode)
    # [:split_offset[*deref_offset2]]
    if HCode[0] == ':':
        hp.type |= USING_SPLIT
        HCode = HCode[1:]
        HCode,hp.split=ConsumeHexInt(HCode)
        if HCode[0] == '*':
            hp.type |= SPLIT_INDIRECT
            HCode = HCode[1:]
            HCode,hp.split_index=ConsumeHexInt(HCode)
    # @addr[:module[:func]]
    match = re.match(r'^@([0-9a-fA-F]+)(:.+?)?(:.+)?$', HCode)
    if match is None:
        return None
    hp.address = int(match.group(1), 16)
    if match.group(2):
        hp.type |= MODULE_OFFSET
        hp.module = match.group(2)[1:]
    if match.group(3):
        hp.type |= FUNCTION_OFFSET
        hp.function =bytes(match.group(3)[1:],encoding='ascii')

    # ITH has registers offset by 4 vs AGTH: need this to correct
    if hp.offset < 0:
        hp.offset -= 4
    if hp.split < 0:
        hp.split -= 4

    return hp 
def Parse(code):
    code=code.strip().replace('\r','').replace('\n','').replace('\t','')
    if(code[0]=='/'):code=code[1:]
    if('/' in code):code=code.split('/')[0]
    if(code[0]=='R'):
        hp=ParseRCode(code[1:])
    elif(code[0]=='H'):
        hp=ParseHCode(code[1:])
    else:
        hp=None
    return hp
def GenerateRCode(hp):
    RCode = 'R'
    if hp.type & USING_UNICODE:
        if hp.type & HEX_DUMP:
            RCode += 'M'
        else:
            RCode += 'Q'
        if hp.null_length != 0:
            RCode += str(hp.null_length) + '<'
    else:
        RCode += 'S'
        if hp.null_length != 0:
            RCode += str(hp.null_length) + '<'
        if hp.codepage != 0:
            RCode += str(hp.codepage) + '#'
    RCode += '@' + Hex(hp.address)
    return RCode

def GenerateHCode(hp,processId):
    HCode = "H"
    if hp.type & USING_UNICODE:
        if hp.type & HEX_DUMP:
            if hp.type & USING_STRING:
                HCode += "M"
            else:
                HCode += "H"
        else:
            if hp.type & USING_STRING:
                HCode += "Q"
            else:
                HCode += "W"
    
    else:
        if hp.type & USING_STRING:
            HCode += "S"
        elif hp.type & BIG_ENDIAN:
            HCode += "A"
        else:
            HCode += "B"
    if hp.type & FULL_STRING:
        HCode += "F"

    if hp.null_length != 0:
        HCode += str(hp.null_length) + "<"
    if hp.type & NO_CONTEXT:
        HCode += "N"
    
    if hp.text_fun or hp.filter_fun or hp.hook_fun or hp.length_fun:
        HCode += "X" # no AGTH equivalent
    if hp.codepage != 0 and not (hp.type & USING_UNICODE):
        HCode += str(hp.codepage) + "#"

    if hp.padding:
        HCode += Hex(hp.padding) + "+"
    if hp.offset < 0:
        hp.offset += 4
    
    if hp.split < 0:
        hp.split += 4
    HCode += Hex(hp.offset)
    if hp.type & DATA_INDIRECT:
        HCode += "*" + Hex(hp.index)
    if hp.type & USING_SPLIT:
        HCode += ":" + Hex(hp.split)
    if hp.type & SPLIT_INDIRECT:
        HCode += "*" + Hex(hp.split_index)
    # Attempt to make the address relative
    try:
        if processId and not (hp.type & MODULE_OFFSET):
            process =windows.AutoHandle(windows.OpenProcess(windows.PROCESS_VM_READ | windows.PROCESS_QUERY_INFORMATION, False, processId))
            if process:
                info=windows.VirtualQueryEx(process,hp.address) 
                if info.AllocationBase:
                    module_name =windows.GetModuleFileNameEx(process, info.AllocationBase)
                    if module_name:
                        hp.address -= info.AllocationBase
                        hp.type |= MODULE_OFFSET
                        hp.module = module_name[module_name.rfind('\\')+1:][:120]
              
    except:
        pass
        #print_exc()
    HCode += "@" + Hex(hp.address)

    if hp.type & MODULE_OFFSET:
        HCode += ":" + hp.module

    if hp.type & FUNCTION_OFFSET:
        HCode += ":" + hp.function.decode('ascii')
    return HCode
def Generate(_hp,process_id):
    hp=copy.copy(_hp)
    if hp.type&DIRECT_READ :
        code=GenerateRCode(hp) 
    else:
        code=GenerateHCode(hp,process_id)
     
    return code 
if __name__=='__main__':
    # print(Parse("/HQN936#1+-c*C:C*1C@4AA:gdi.dll:GetTextOutA",hp))
    # print(Parse("/HQN936#-c*C:C*1C@4AA:gdi.dll:GetTextOutA /KF",hp))
    # print(Parse("HB4@0" ,hp)),
    # print(Parse("/RS65001#@44",hp)),
    # print(Parse("HQ@4",hp,))
    print(Parse('/HS8:-14@76D85270'))
    # print(Parse("/RW@44",hp)),
    # print(Parse("/HWG@33",hp))
    